Source/ExecutionEngine/ExecutionEngine.cs

```diff
@@ -882,15 +827,53 @@ private Implementation[] GetPrioritizedImplementations(Program program)
       }
 
       if (Options.PrintNecessaryAssumes && program.NecessaryAssumes.Any()) {
-        Console.WriteLine("Necessary assume command(s): {0}", string.Join(", ", program.NecessaryAssumes));
+        Console.WriteLine("Necessary assume command(s): {0}", string.Join(", ", program.NecessaryAssumes.OrderBy(s => s)));
       }
 
       cce.NonNull(Options.TheProverFactory).Close();
 
       return outcome;
+
+    }
+
+    async Task<VerificationResult> VerifyImplementationWithLargeStackScheduler(
+      Program program, PipelineStatistics stats,
+      string programId, ErrorReporterDelegate er, string requestId, Implementation[] stablePrioritizedImpls,
+      Dictionary<string, Dictionary<string, Block>> extractLoopMappingInfo,
+      CancellationTokenSource cts,
+      int index, TextWriter taskWriter)
+    {
+      var implementation = stablePrioritizedImpls[index];
+      var id = implementation.Id;
+      if (ImplIdToCancellationTokenSource.TryGetValue(id, out var old)) {
+        old.Cancel();
+      }
+
+      await verifyImplementationSemaphore.WaitAsync(cts.Token);
+      try {
+
+        ImplIdToCancellationTokenSource.AddOrUpdate(id, cts, (k, ov) => cts);
+
+        var coreTask = new Task<VerificationResult>(() => VerifyImplementation(program, stats, er, requestId,
+          extractLoopMappingInfo, implementation,
+          programId, taskWriter).Result, cts.Token, TaskCreationOptions.None);
```

Comment:
> Why not directly await VerifyImplementation? Is it because you need a different cancellation token?

Because then you can't use a custom thread scheduler.

> Also, I think you can use [TaskFactory.StartNew](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskfactory-1.startnew?view=net-6.0#system-threading-tasks-taskfactory-1-startnew(system-func((system-object-0))-system-object-system-threading-cancellationtoken-system-threading-tasks-taskcreationoptions-system-threading-tasks-taskscheduler)) directly instead of new Task + .Start, which is at least what the documentation recomments.

That looks nicer yes, but note that I'm changing this line again a little bit in https://github.com/boogie-org/boogie/pull/538/files/b0214fc8dd58f62c3acf88c8eb914fc05c26be87..6a9a3a7b4c3d757b1af78cd666952878dec8a028#diff-a5b39418ae8731b4b74a8cb24041d0272c8e4a02316c312cdc2c1bb19dff7d02R857

And I want to change it a little more later to `Task.Run(() => VerifyImplemen...)` once we make sure we don't need the large stack thread scheduler: https://github.com/boogie-org/boogie/pull/524

> I'm not 100% sure what's going on here: you're forcing the task VerifyImplementation to run synchronously by taking its result, but the whole thing is still asynchronous because of the new Task around it?

Yes but it does cause the created thread to wait during IO operations so it's not a great use of resources, that's one thing https://github.com/boogie-org/boogie/pull/524 improves.

Reply:
Oh, I see.  Then I guess this is fine and we can leave future improvements to later.